File: go/math/149_max_points.go
package math

// maxPoints 直线上最多的点数
func maxPoints(points [][]int) int {
	res := 0

	for i := 0; i < len(points); i++ {
		// 建立一个key为两点的斜率，可能是浮点数
		hmap := make(map[float64]int)

		for j := 0; j < len(points); j++ {
			// 不能是同一个点组成的直线
			if i != j {
				hmap[lineSlop(points[i], points[j])]++
			}
		}

		for _, v := range hmap {
			if v > res {
				res = v
			}
		}
	}

	return res + 1 // hash表中没有统计点i 自身，所以需要加1
}

// lineSlop 斜率 float 计算除数为0 的时候，会转为inf
func lineSlop(a, b []int) float64 {
	return float64(a[1]-b[1]) / float64(a[0]-b[0]) // (ay-by)/(ax-bx)
}

File: rust/src/math/_149_max_points.rs
/// 直线上最多的点数
pub fn max_points(points: Vec<Vec<i32>>) -> i32 {
    let mut res = 0;

    for i in 0..points.len() {
        use std::collections::HashMap;
        let mut hmap = HashMap::new();

        for j in 0..points.len() {
            if i != j {
                hmap.entry(line_scop(&points[i], &points[j]))
                    .and_modify(|x| *x += 1)
                    .or_insert(1);
            }
        }

        // 遍历完之后，比较哪个值大
        for (_, v) in hmap {
            if v > res {
                res = v
            }
        }
    }

    res + 1
}

fn line_scop(a: &[i32], b: &[i32]) -> (i32, i32) {
    let dx = a[0] - b[0];
    let dy = a[1] - b[1];

    if dx == 0 {
        return (0, 1); // 垂直线，表示斜率为正无穷大
    }

    let gcd_val = gcd(dx.abs(), dy.abs());
    ((dx / gcd_val), dy / gcd_val)
}

fn gcd(mut a: i32, mut b: i32) -> i32 {
    while b != 0 {
        let temp = b;
        b = a % b;
        a = temp;
    }
    a
}
